content/post/debrick-core9g25-intro.md

```diff
@@ -0,0 +1,131 @@
+
+#### Read 0x5C42(23618) bytes ###
+Erase NAND flash 0x180000 0x80000 OK
+
+########## Reading uImage ###########
+####################
+#### Read 0x304FDA(3166170) bytes ###
+Erase NAND flash 0x200000 0x600000 OK
+
+########## Reading rootfs.img ###########
+####################
+#### Read 0xE40000(14942208) bytes ###
+Erase NAND flash 0x800000 0xf800000 OK
+
+### Update OK ### Please RESET the board ###
+```
+
+As you can see U-Boot updates the NAND using 5 files.
+
+* `at91sam9x5ek-nandflashboot-uboot-3.5.3.bin` - the bootloader
+[AT91Bootstrap][at91bootstrap]
+* `u-boot.bin` - the bootloader U-Boot
+* `at91sam9g25ek.dtb` - the device tree blob
+* `uImage` - the Linux kernel
+* `rootfs.img` - the root filesystem
+
+First U-Boot loads the file `at91sam9x5ek-nandflashboot-uboot-3.5.3.bin`, than
+it ereases a block of `0x40000` bytes on the NAND, starting at `0x0`. It then
+writes the file to Nand, aligning the program at address `0x0`.
+
+## Risks
+Because U-Boot performs the update, updating U-Boot is risky.  What happens if
+the binaries for AT91Boostrap or U-Boot aren't compiled correctly? Or what if
+the power is lost during the update?  Or what if you've replaced the original
+U-Boot with a version that doesn't contain the update script (yes, I've done
+that)? Well, you're devices bricked.
+
+Only recently together my college Marten van Houten and I managed to reflash
+the NAND using tools provided but the vendor of the CPU.
+
+In order to understand fully how we managed to do that I'll explain the boot
+process of AT91SAM9G25 CPU.
+
+## Boot process
+The following snippet is taken from chapter 10 'Boot strategies' of [AT91SAM9G25's
+datasheet][datasheet] and it explains the boot process:
+
+> The system always boots at address `0x0`. To ensure maximum boot
+> possibilities, the memory layout can be changed with the BMS pin. This allows
+> the user to lay out the ROM or an external memory to `0x0`. The sampling of
+> the BMS pin is done at reset. If BMS is detected at `0`, the controller boots
+> on the memory connected to Chip Select `0` of the External Bus.
+>
+> [...]
+>
+> If BMS is detected at `1`, the boot memory is the embedded ROM and the Boot
+> Program described below is executed.
+
+Upon receiving power the CPU starts the program at address `0x0`.  Normally,
+that is when the BMS pin is `0`, the NAND is layed out to this address. So
+when the CPU starts the program that is located at `0x0` on the NAND.
+
+And which program is at this address? Well, that is AT91Bootstrap.
+
+## SAM-Ba to the rescue
+But if you set the BMS pin to `1`, the memory lay out is modified.
+Instead of the NAND, the ROM is now layed out at address `0x0`. The program
+that is found on `0x0` on the ROM will be executed.
+
+The AT91SAM9G25 CPU comes with a 64 kB ROM. This ROM contains a program called
+[SAM boot agent (SAM-BA)][sam-ba]. This program allows one to reprogram the
+NAND flash. But I'll discuss that in 2 other posts.
```

Comment:
This post is ended all of a sudden.  It is not that clear when the boot proces has ended. I assume when `The program that is found on `0x0` on the ROM` has been executed? If so, you could add `And that's the boot proces of this AT91SAM9G25 CPU`

For the last paragraph, remove the first sentence and start with `This ROM (64kB) contains..` `... and this program allows one...` (one sentence).